source/kraken/fill_disruption_from_chaos.cpp

```diff
@@ -443,79 +717,113 @@ struct add_impacts_visitor : public apply_impacts_visitor {
-        this->log_start_action(stop_point->uri);
-        if (stop_point->journey_pattern_point_list.empty()) {
+    void remove_jp_if_no_vj(nt::JourneyPattern* jp, const std::set<nt::new_disruption::Impact*>& jp_impacts) {
+
+        // test if there are any vj still attched to the jp
+        if(! jp->discrete_vehicle_journey_list.empty() ||
+                ! jp->frequency_vehicle_journey_list.empty()) {
             return;
         }
-        // We copy this journey_pattern
-        for (const auto jpp_ref : stop_point->journey_pattern_point_list) {
-            const auto jp_ref = jpp_ref->journey_pattern;
-            auto new_jp = new nt::JourneyPattern(*jp_ref);
-            new_jp->uri = make_adapted_uri_fast(new_jp->uri, pt_data.journey_patterns.size());
-            new_jp->idx = pt_data.journey_patterns.size();
-            pt_data.journey_patterns.push_back(new_jp);
-            pt_data.journey_patterns_map[new_jp->uri] = new_jp;
-            new_jp->journey_pattern_point_list.clear();
-            new_jp->discrete_vehicle_journey_list.clear();
-            new_jp->frequency_vehicle_journey_list.clear();
-            // We copy each journey_pattern_point but the ones which are linked
-            // to the impacted stop_area
-            size_t order = 0;
-            for (const auto jpp_to_copy : jp_ref->journey_pattern_point_list) {
-                if (jpp_to_copy->stop_point == stop_point) {
-                    continue;
-                }
-                auto new_jpp = new nt::JourneyPatternPoint(*jpp_to_copy);
-                new_jpp->idx = pt_data.journey_pattern_points.size();
-                new_jpp->uri = make_adapted_uri_fast(new_jpp->uri, pt_data.journey_pattern_points.size());
-                new_jp->journey_pattern_point_list.push_back(new_jpp);
-                new_jpp->stop_point->journey_pattern_point_list.push_back(new_jpp);
-                pt_data.journey_pattern_points.push_back(new_jpp);
-                pt_data.journey_pattern_points_map[new_jpp->uri] = new_jpp;
-                new_jpp->journey_pattern = new_jp;
-                new_jpp->order = order;
-                ++ order;
-            }
+        // let's take advantage of RAII of unique_ptr
+        auto jp_unique_ptr = std::unique_ptr<nt::JourneyPattern>{jp};
+
+        // first, we remove all jpp in jp
+        for(auto* jpp: jp->journey_pattern_point_list){
+            // let's take advantage of RAII of unique_ptr
+            auto jpp_unique_ptr = std::unique_ptr<nt::JourneyPatternPoint>{jpp};
 
-            // We copy the vehicle_journeys of the journey_pattern_points
-            functor_add_vj v(impact, new_jp, stop_point, pt_data, meta);
-            jp_ref->for_each_vehicle_journey(v);
-            // The new journey_pattern is linked to the impact
-            impact->impacted_journey_patterns.push_back(new_jp);
+            erase_remove_helper(pt_data.journey_pattern_points, jpp);
```

Comment:
that'll be costy as you'll move every jpps for each jpp in the jp... Better do that in one pass.